src/boot/linux.c

```diff
@@ -217,67 +227,51 @@ EFI_STATUS linux_exec(
-
-        *loaded_image = (EFI_LOADED_IMAGE_PROTOCOL) {
-                .Revision = 0x1000,
-                .ParentHandle = parent,
-                .SystemTable = ST,
-                .DeviceHandle = parent_loaded_image->DeviceHandle,
-                .FilePath = file_path,
-                .ImageBase = loaded_kernel,
-                .ImageSize = kernel_size_in_memory,
-                .ImageCodeType = 1 /* EFI_LOADER_CODE */,
-                .ImageDataType = 2 /* EFI_LOADER_DATA */,
-        };
+        parent_loaded_image->ImageBase = loaded_kernel;
+        parent_loaded_image->ImageSize = kernel_size_in_memory;
 
         if (cmdline) {
-                loaded_image->LoadOptions = (void *) cmdline;
-                loaded_image->LoadOptionsSize = strsize16(loaded_image->LoadOptions);
+                parent_loaded_image->LoadOptions = (void *) cmdline;
+                parent_loaded_image->LoadOptionsSize = strsize16(parent_loaded_image->LoadOptions);
         }
 
         _cleanup_(cleanup_initrd) EFI_HANDLE initrd_handle = NULL;
         err = initrd_register(initrd->iov_base, initrd->iov_len, &initrd_handle);
         if (err != EFI_SUCCESS)
                 return log_error_status(err, "Error registering initrd: %m");
 
-        EFI_HANDLE kernel_image = NULL;
-
-        err = BS->InstallMultipleProtocolInterfaces(
-                        &kernel_image, MAKE_GUID_PTR(EFI_LOADED_IMAGE_PROTOCOL), loaded_image,
-                        NULL);
-        if (err != EFI_SUCCESS)
-                return log_error_status(err, "Cannot install loaded image protocol: %m");
-
         log_wait();
 
         if (entry_point > 0) {
                 EFI_IMAGE_ENTRY_POINT entry =
-                        (EFI_IMAGE_ENTRY_POINT) ((const uint8_t *) loaded_image->ImageBase + entry_point);
-                err = entry(kernel_image, ST);
+                        (EFI_IMAGE_ENTRY_POINT) ((const uint8_t *) parent_loaded_image->ImageBase + entry_point);
+                err = entry(parent_image, ST);
         } else if (compat_entry_point > 0) {
                 /* Try calling the kernel compat entry point if one exists. */
                 EFI_IMAGE_ENTRY_POINT compat_entry =
-                                (EFI_IMAGE_ENTRY_POINT) ((const uint8_t *) loaded_image->ImageBase + compat_entry_point);
-                err = compat_entry(kernel_image, ST);
+                                (EFI_IMAGE_ENTRY_POINT) ((const uint8_t *) parent_loaded_image->ImageBase + compat_entry_point);
+                err = compat_entry(parent_image, ST);
         }
 
-        EFI_STATUS uninstall_err = BS->UninstallMultipleProtocolInterfaces(
-                        kernel_image, MAKE_GUID_PTR(EFI_LOADED_IMAGE_PROTOCOL), loaded_image,
-                        NULL);
-        if (uninstall_err != EFI_SUCCESS)
-                return log_error_status(uninstall_err, "Cannot uninstall loaded image protocol: %m");
```

Comment:
hmm, shouldn't we patch the old fields back in on failure? in particular the ones pointing to garbage now?